### ISSUE TITLE
Poor training results

### ISSUE THREAD

**Issue by Rizhiy** [roytseng-tw/Detectron.pytorch#24]
Hi, I have trained R-101-FPN with coco2017, using 4 GPUs, but only got mmAP=0.33 during test which is well below Detectron result of 0.40.

What can be the problem?

I have used `python tools/train_net.py --dataset coco2017 --cfg configs/e2e_mask_rcnn_R-101-FPN_2x.yaml --use-tfboard --nw 8 --b 8` for training and `python tools/test_net.py --cfg configs/e2e_mask_rcnn_R-101-FPN_2x.yaml --load_ckpt Outputs/e2e_mask_rcnn_R-101-FPN_2x/Apr19-11-34-35_devbox/ckpt/model_7_29315.pth --dataset coco2017`

The loss at the end was about 0.6 which also seems a bit high.

**Comment by roytseng-tw** [comment]
Hello, apparently you didn't adjust the training config (training epochs, learning rate decay epochs) to match the official one.
I recently update a training script `tools/train_net_step.py`, which works with `SOLVER.LR_POLICY: steps_with_decay` and `SOLVER.WARM_UP_METHOD = 'linear'`(default). By using this training script, one should totally replicate the training procedures that are consistent with the above two config options from Detectron. I already retrain a `e2e_mask_rcnn_R-50-FPN_1x.yaml` from scratch for a few days, and the training should end in 1 or 2 more days. Every training settings are kept the same, including the warm-up iterations, except that I only use 4 gpus (half numbers of the default) for same batch size of 16.
Hope I'll have good news for you later on ! ><

**Comment by Rizhiy** [comment]
Thanks, can you tell me exactly what should I adjust in the config to reproduce the result? It is not very clear in the `README.md`.

**Comment by roytseng-tw** [comment]
It's enough to just run `python tools/train_net_step.py --dataset coco2017 --cfg e2e_mask_rcnn_R-50-FPN_1x.yaml --use-tfboard --nw 10`  The batch size should be 16, so adjust number of workers as you wish.
BTW, I just found a bug in this script about decaying the learning rate. Remember to pull the newest one.

**Comment by Rizhiy** [comment]
Hi I tried to run `python tools/train_net_step.py --dataset coco2017 --cfg e2e_mask_rcnn_R-50-FPN_1x.yaml --use-tfboard --nw 8` on commit 162c11e and only got 0.3 mmAP.

**Comment by Rizhiy** [comment]
Ok, I have put on training with 4 GPUs using latest commit (7d3cf1e), will update in about 3 days.

**Comment by roytseng-tw** [comment]
Hi, I have updated the benchmark results trained with 4 GPUs, 16 batch size on `e2e_mask_rcnn_R-50-FPN_1x`. The scores are very close to the official ones (drop < 0.5%)

**Comment by Rizhiy** [comment]
I have finished training using 4 GPUS and got 0.377 box mmAP and 0.340 mask mmAP. So should be correct now.

**Comment by fitsumreda** [comment]
@Rizhiy , could you share which commit and command you used to reproduce those results?
also, how many gpus did you use?
tahnks!
cc @roytseng-tw

**Comment by roytseng-tw** [comment]
@Rizhiy,  Did you change `TRAIN.RPN_PRE_NMS_TOP_N` and `TRAIN.RPN_POST_NMS_TOP_N` yourself ? I found it necessary to reproduce the accuracy of Detectron, if number of batches per GPU 
, namely `TRAIN.IMS_PER_BATCH`, is different from the original config. 

For your case, you use 4 GPUs insteads of 8, so `TRAIN.IMS_PER_BATCH` is doubled and you should also double the value of `TRAIN.RPN_PRE_NMS_TOP_N` and `TRAIN.RPN_POST_NMS_TOP_N`. BTW, this adjustment has been implemented in train_net_step.py

**Comment by Rizhiy** [comment]
@fitsumreda I run `python tools/train_net_step.py --dataset coco2017 --cfg e2e_mask_rcnn_R-50-FPN_1x.yaml --use-tfboard --nw 8`, can't say exactly which I commit I used.

@roytseng-tw No, I didn't change any settings, just used the config. I didn't alter `TRAIN.IMS_PER_BATCH` either.
TBH `TRAIN.RPN_PRE_NMS_TOP_N` and `TRAIN.RPN_POST_NMS_TOP_N` should be per img, otherwise the code is rather bad.

**Comment by Rizhiy** [comment]
@roytseng-tw You are creating `cfg.FPN.RPN_COLLECT_SCALE` directly during training and currently it is only being created in one of the scripts. I think this is bad practice, you should probably add it to base config.

**Comment by Rizhiy** [comment]
Actually, it looks like you need to adjust `cfg.TRAIN.RPN_PRE_NMS_TOP_N`
as well.

**Comment by roytseng-tw** [comment]
@Rizhiy can you share your training log and the final checkpoint to me ? I still can't get mask AP to 34 on `e2e_mask_rcnn_R-50-FPN_1x.yaml`, and I want to compare the losses. BTW, is it correct that you were using 4 GPUs to do the training ?

**Comment by fitsumreda** [comment]
@roytseng-tw @Rizhiy 
fyi, I can now reproduce `configs/e2e_mask_rcnn_R-50-FPN_1x.yaml`
I used a batch size of 16, 8 V100 GPUs, 0.02 learning rate, and 180K iterations (SOLVER.STEPS scaled accordingly).  Every other parameters stay the same as defined in the config yaml file.
thanks!

**Comment by roytseng-tw** [comment]
@fitsumreda
Wonderful ~ Glad to hear that !

**Comment by karenyun** [comment]
@roytseng-tw 
Hi, thanks for the great code!
I have a question about the above you have discussed: you mean the "batch_size" should be 16 as the official rather the "effective batch_size"?

**Comment by qinhaifangpku** [comment]
> @roytseng-tw @Rizhiy
> fyi, I can now reproduce `configs/e2e_mask_rcnn_R-50-FPN_1x.yaml`
> I used a batch size of 16, 8 V100 GPUs, 0.02 learning rate, and 180K iterations (SOLVER.STEPS scaled accordingly). All other parameters stay the same as defined in the config yaml file.
> thanks!

@fitsumreda Hi, I am now dound the code could not run on v100, cuda9.0

could you please help me out of the problem?

**cudaCheckError() failed : no kernel image is available for execution on the device
THCudaCheck FAIL file=torch/csrc/cuda/Module.cpp line=32 error=29 : driver shutting down
Segmentation fault**
